lib/ex_unit/lib/ex_unit/manifest.ex

```diff
@@ -4,19 +4,41 @@ defmodule ExUnit.Manifest do
   import Record
 
   defrecord :entry, [:last_run_status, :file]
-  @opaque t :: [{test_id, entry}]
+  defstruct entries: [], path: nil
+
+  @opaque entries :: [{test_id, entry}]
+  @type t :: %__MODULE__{entries: entries, path: Path.t()}
   @type status :: :passed | :failed
   @type entry :: record(:entry, last_run_status: status, file: Path.t())
   @type test_id :: {module, name :: atom}
+  @type last_run_status_index :: %{test_id => status}
 
   @manifest_vsn 1
 
-  @spec new() :: t
-  def new do
-    []
+  @spec new(Path.t()) :: t
+  def new(path) when is_binary(path) do
+    %__MODULE__{path: path}
+  end
+
+  @spec to_last_run_status_index(t) :: last_run_status_index
+  def to_last_run_status_index(%__MODULE__{entries: entries}) do
+    Map.new(entries, fn {test_id, entry(last_run_status: status)} ->
+      {test_id, status}
+    end)
+  end
+
+  @spec get_files_with_failures(t) :: MapSet.t(Path.t())
+  def get_files_with_failures(%__MODULE__{entries: entries}) do
+    entries
+    |> Stream.filter(fn {_, entry(last_run_status: status)} -> status == :failed end)
+    |> MapSet.new(fn {_, entry(file: file)} -> file end)
   end
 
   @spec add_test(t, ExUnit.Test.t()) :: t
+  def add_test(manifest, %ExUnit.Test{tags: %{file: file}})
+      when not is_binary(file),
+      do: manifest
```

Comment:
It's unfortunate this function clause is needed, but without it, `ex_unit_test.exs` experiences order-dependent failures, due to [this test](https://github.com/elixir-lang/elixir/blob/36395a248cfad5bef73f5700785bb575ce34add4/lib/ex_unit/test/ex_unit_test.exs#L381-L400). Apparently when you have a `setup` block like `setup do: {:ok, file: :foo}` it still updates the test's `:file` tag even though it also triggers an exception telling the user the mistake.  The `file: :foo` tag would then get stored in the manifest as the file (as we always keep the results from the new manifest).  Then, on the next test run, when `ExUnit.Manifest.merge/2` is inspecting the old manifest entries, it would try to check `File.regular?(:foo)` and get an exception.  This function clause ignores the test entirely when it's `:file` tag is invalid, working around the problem.

That said, while this fixes the exception, I think a problem still remains.  If a user did `setup do: {:ok, file: "string"}` we would store the test in the manifest with the wrong `:file` value.  I haven't spent any time looking into it yet, but I think a better fix would be to prevent the tag from updating in this case so that the `file` passed to this function is always the real file the test came from.